\section{Introduction}
A local alignment of two similar nucleotide sequences can be
approximated as long exact matches (\ty{anchors}) separated by
mismatches. The use of enhanced suffix arrays (ESA) for finding the
exact matches makes such an approximation fast compared to an actual
alignment. The anchor-based comparison of whole genomes was
implemented in a number of programs
including \ty{andi}~\cite{hau15:and} \ty{phylonium}~\cite{kloe19:phy},
which are used for fast computation of evolutionary distances. In
addition, the method is used for finding intersections between
multiple target genomes in the program \texttt{fur}~\cite{hau21:fur}.

The package \ty{ancs} contains functions for anchor-based
approximation of local alignments.
\newpage
